File: daily_python/3264_Final_Array_State_After_K_Multiplication_Operations_I.py
import os
import sys
import heapq
from typing import List

class Solution:
    def getFinalState(self, nums: List[int], k: int, multiplier: int) -> List[int]:
        n = len(nums)
        minheap = []
        for i, num in enumerate(nums):
            minheap.append((num, i))
        heapq.heapify(minheap)
        for i in range(k):
            curr, index = heapq.heappop(minheap)
            heapq.heappush(minheap, (curr * multiplier, index))
        ans = [0] * n
        while minheap:
            curr, index = heapq.heappop(minheap)
            ans[index] = curr
        return ans
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            nums = list(map(int, lines[0].split('=')[1].strip()[1:-2].split(',')))
            k = int(lines[1].split('=')[1].strip()[:-1])
            multiplier = int(lines[2].split('=')[1].strip()[:-1])
            dataset.append((nums, k, multiplier))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for nums, k, multiplier in dataset:
        results.append(solution.getFinalState(nums, k, multiplier))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')
    

File: daily_python/3264_Final_Array_State_After_K_Multiplication_Operations_I.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

class Solution {
public:
    vector<int> getFinalState(vector<int>& nums, int k, int multiplier) {
        auto cmp = [&](int a, int b){
            if (nums[a] == nums[b]) return a > b;
            else return nums[a] > nums[b];
        };
        priority_queue<int, vector<int>, decltype(cmp)> pq(cmp);
        int n = nums.size();
        for (int i = 0; i < n; i++) pq.push(i);
        while(k--){
            int i = pq.top();
            pq.pop();
            nums[i] *= multiplier;
            pq.push(i);
        }
        return nums;
    }
};

int main(){
    vector<int> nums = {2,1,3,5,6};
    int k = 5;
    int multiplier = 2;
    // vector<int> nums = {1,2};
    // int k = 3;
    // int multiplier = 4;
    Solution S;
    vector<int> ans = S.getFinalState(nums, k, multiplier);
    for (int i : ans) cout << i << " ";
    cout << endl;
    return 0;
}